src/main/java/io/cryostat/recordings/ArchiveRequestGenerator.java

```diff
@@ -0,0 +1,278 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.cryostat.recordings;
+
+import java.time.Duration;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+import io.cryostat.ConfigProperties;
+import io.cryostat.core.reports.InterruptibleReportGenerator.AnalysisResult;
+import io.cryostat.reports.ReportsService;
+import io.cryostat.ws.MessagingServer;
+import io.cryostat.ws.Notification;
+
+import io.quarkus.vertx.ConsumeEvent;
+import io.vertx.mutiny.core.eventbus.EventBus;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import org.apache.commons.lang3.tuple.Pair;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ApplicationScoped
+public class ArchiveRequestGenerator {
+
+    public static final String ARCHIVE_ADDRESS =
+            "io.cryostat.recordings.ArchiveRequestGenerator.ArchiveRequest";
+    public static final String GRAFANA_ARCHIVE_ADDRESS =
+            "io.cryostat.recordings.ArchiveRequestGenerator.GrafanaArchiveUploadRequest";
+    public static final String GRAFANA_ACTIVE_ADDRESS =
+            "io.cryostat.recordings.ArchiveRequestGenerator.GrafanaActiveUploadRequest";
+    public static final String ARCHIVE_REPORT_ADDRESS =
+            "io.cryostat.recordings.ArchiveRequestGenerator.ArchiveReportRequest";
+    public static final String ACTIVE_REPORT_ADDRESS =
+            "io.cryostat.recordings.ArchiveRequestGenerator.ActiveReportRequest";
+    private static final String ARCHIVE_RECORDING_SUCCESS = "ArchiveRecordingSuccess";
+    private static final String ARCHIVE_RECORDING_FAIL = "ArchiveRecordingFailed";
+    private static final String GRAFANA_UPLOAD_SUCCESS = "GrafanaUploadSuccess";
+    private static final String GRAFANA_UPLOAD_FAIL = "GrafanaUploadFailed";
+    private static final String REPORT_SUCCESS = "ReportSuccess";
+    private static final String REPORT_FAILURE = "ReportFailure";
+    private final ExecutorService executor;
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Inject private EventBus bus;
+    @Inject private RecordingHelper recordingHelper;
+    @Inject ReportsService reportsService;
+
+    private Map<String, Map<String, AnalysisResult>> jobResults;
```

Comment:
I would prefer to avoid adding something like this that hangs on to the report generation results in memory. The report generation system already has tiered in-memory and S3-backed caching, so it'd be best to lean on that.

For archiving recordings, or for uploading recording files to jfr-datasource, it's a simpler design because there is no response payload to send back to the client - the client just needs to know that the job has been completed and whether it completed successfully.

For report generation, where it takes time and the client needs to not only not that it has completed and whether it was a success, but also needs a response payload (report result), I think an API that behaves something like `Map.computeIfAbsent()` would be best. I'm thinking that the `ReportsService` interface should gain methods for checking whether a result is already present for a given key. The default implementation would trivially return false, and the two caching tier implementations would check their respective caches and delegates.

![image](https://github.com/user-attachments/assets/7f344fb2-7257-4462-bb0e-8d2b40e1273c)

HTTP 200 responses would contain the `Map<String, AnalysisResult>` as the response body for the client to use. HTTP 202 responses would contain a Job UUID in the response body, maybe a `Location` header to tell the client the URL where to find the result when it's ready (1), and that's it. Later, the completion notification containing the Job UUID would be emitted. When the client receives the the notification containing that UUID it sends a follow-up GET to the `Location` header URL, and this time around the result is ready to go so the response is an HTTP 200 with the expected data. This way, the existing tiered caching infrastructure is reused and there are no new endpoints introduced, only different response status behaviours on the existing endpoints.

(1) right now, the URL would be the same URL that the request was sent to initially, ex. `/api/v4/reports/{encodedKey}`

Reply:
Added a WIP implementation of the suggested design, returning a 200 with the analysis result if the storage or quarkus cache has the report, delegating to the generator otherwise. The web client needs to be updated to react to the notifications which is what I'll work on next, but for now it gets the 202 response when the cache misses with the location URL to check when it gets the notification